### PR TITLE
⚡️ perf(curveframes): amortise arc length with an optional s_max precompute

### LINKED ISSUE #713: Amortise the arc-length reparametrisation without breaking gradients
`ArcLength` (added in #712) reparametrises a curve by arc length, solving `dτ/ds = 1/‖γ′(τ)‖` on **every call**. Under `BishopBuilder` that sits inside Bishop's own parallel-transport solve, which evaluates the curve many times per call, so the costs multiply.

Measured on a helix, float64, warmed up, steady state:

| | n=1 | n=100 |
|---|---|---|
| plain forward | 129.3 µs | 10.7 µs/pt |
| **arc-length forward** | **5976.2 µs** | **343.0 µs/pt** |
| plain gradient | 1724.9 µs | |
| **arc-length gradient** | **148899.2 µs** | |

**46× forward, 86× under `grad`** — 149 ms per gradient call. Frenet–Serret is far cheaper, having no ODE of its own.

## What was tried, and why it was reverted

An optional `s_max` precomputed τ(s) once into a `diffrax` dense interpolation. It worked, and it was fast: forward dropped from 6183 µs to 806 µs, a **7.7×** speedup, with the two paths agreeing to 2.22e-16.

It was reverted because **it returned silently wrong gradients**.

The interpolation was built in `__init__`. When the module is constructed *outside* the differentiated function — the amortising pattern the feature exists for — perturbing the curve's leaves does not rebuild it. The τ(s) map stays frozen at the original parameters, and only the final `curve(τ)` evaluation sees the perturbation:

```
plain  d/dpitch = 1.7574794224     (finite difference agrees)
cached d/dpitch = 1.9156525704     wrong
```

For `LagrangianArcLength.t0` on a bending curve, AD returned **exactly 0.0** against a finite difference of −0.3435633403.

Building the module *inside* the trace gives correct gradients — but that is precisely when the precompute buys nothing.

## Why this is structural

Any precompute stored as a pytree leaf is stale with respect to perturbations of its inputs. It is not a bug to patch: the derivative of the interpolation w.r.t. the curve's parameters is genuinely missing, not merely inaccurate. Recomputing inside `__call__` restores correctness and removes the speedup.

## What a real fix needs

`jax.custom_vjp` around the reparametrisation, supplying the correct derivative analytically instead of differentiating through a stale artefact. From `s = ∫₀^τ ‖γ′(u)‖ du`, the sensitivity of τ(s) to a curve parameter θ is obtainable by implicit differentiation:

    ∂τ/∂θ = −(∂s/∂θ)|_τ / ‖γ′(τ)‖

so the correct VJP needs one extra quadrature of `∂‖γ′‖/∂θ` along the interval, not a re-solve of the whole ODE. That should be cheap enough to keep most of the 7.7×.

## Acceptance criteria

Whatever the approach:

1. **Gradients w.r.t. curve parameters and `t0` must match finite differences** with the module built *outside* the differentiated function. That is the case that broke; a test differentiating only w.r.t. `s` gave false assurance and must not be the guard.
2. The fast and slow paths must agree to ~1e-15 on a **non-constant-speed** curve. A constant-speed helix cannot distinguish them.
3. It must work inside `TubularChart`. The reverted version's domain guard had no tolerance, so a legitimate `nearest_tau` result of `s = −4.73e-9` at a chart's `tau = 0` endpoint raised — 4 of 5 round-trips failed with `s_max` set equal to `tau_bounds`. **No test combined the precompute with a chart**, which is how that reached final review.
4. Interaction with `TubularChart.tau_bounds` documented — they are the same quantity at two layers and can disagree.

Reverted in #712; the guide states the cost plainly meanwhile.

### PR DESCRIPTION
**Pull request by nstarman**
Closes #713.

`ArcLength`/`LagrangianArcLength` gain an optional `s_max`. Given one, the reparametrisation ODE is solved once in `__init__` over `[0, s_max]` with `SaveAt(dense=True)`, and later calls interpolate rather than re-solving.

## Why the previous attempt failed

#712 implemented the same caching and reverted it: gradients came back silently wrong, because autodiff ran through a `_interp` built from the *old* curve parameters. It returned `1.9156525704` where the truth is `-1.7574794224` — the *primal* tau value, not its derivative — and exactly `0.0` for `t0`.

The fix is to not differentiate the cache at all, but to supply the derivative analytically, from the implicit function theorem:

```
dtau/dtheta|_s = -(dS/dtheta)|_tau / ||gamma'(tau)||
```

verified to 1.24e-10 against central differences on a helix.

## Correction to the issue: custom_jvp, not custom_vjp

#713 specified a `custom_vjp`. That is the wrong primitive here, and I have [commented on the issue](https://github.com/GalacticDynamics/coordinax/issues/713#issuecomment-5294908438) to that effect.

`BishopBuilder` differentiates *forward* through the curve — `jacfwd` at `bishop.py:338` and again at `:377` — and a `custom_vjp` cannot be `jvp`-ed. This repo already documents the hazard at `bishop.py:68`, for `RecursiveCheckpointAdjoint`, for the same reason. Demonstrated minimally:

| | `grad` | `jacfwd` |
|---|---|---|
| `custom_vjp` | 0.540302 | `TypeError: can't apply forward-mode autodiff (jvp) to a custom_vjp function` |
| `custom_jvp` | 0.540302 | 0.540302 |

So the rule is an `eqx.filter_custom_jvp`: forward mode directly, reverse mode by transposition. The inner `eqx.filter_jvp` of the quadrature needs `DirectAdjoint` for the same reason.

## What it speeds up, and what it does not

Measured on a helix, eager, `BishopBuilder.location`:

| | penalty vs raw curve | with `s_max` | recovered |
|---|---|---|---|
| forward | 941x | 3.8x | **248x** |
| gradient | 208x | 114x | **1.8x** |

**The forward path is amortised; gradients are essentially not.** The JVP rule still integrates `dS/dtheta` over `[tau_0, tau]` on every call, so only the primal solve is cached. Since part of #713's motivation was gradient-based fitting, that gap is worth knowing about before this is treated as closing the performance question. The docs say so plainly rather than quoting one headline number.

Under `eqx.filter_jit` neither shows a difference at all: per-call cost there is dominated by dispatch overhead (#719), which swamps the ODE entirely.

## Domain handling

`_eval_tau_dense` clamps within `_S_MAX_MARGIN` (5%) of the domain instead of gating exactly on `[0, s_max]`. `nearest_tau`'s bracketed root-find structurally probes a full scan-seed spacing outside `tau_bounds`, so an exact gate raises on legitimate converged results — #712 measured one at `s = -4.73e-9`. Beyond the margin it raises rather than letting `diffrax` return `NaN` silently.

One consequence worth a reviewer's eye: a query up to 5% past `s_max` is silently clamped and returns a confidently wrong position. The margin is sized to the root-find's structural need, but it does mean genuine caller error in that band is not reported.

`s_max` is rejected for a two-argument curve — the Eulerian reading re-measures arc length per slice, so no single tau(s) exists to precompute. Bind the slice with `AtTime` first. `LagrangianArcLength` always accepts it, since `t0` is fixed by construction.

## Verification

Every test was checked to actually fail, by breaking the implementation:

| mutation | result |
|---|---|
| JVP formula scaled by 2 | all 5 gradient tests fail, including the Bishop one |
| `_S_MAX_MARGIN = 0.0` | chart boundary test fails with `EquinoxRuntimeError` at `nearest.py:142` — the reverted PR's exact bug |
| `custom_vjp` under `jacfwd` | `TypeError`, as above |

Gradients are verified with the module built **outside** the differentiated function (via `eqx.tree_at`), which is the pattern that broke before, and the precompute is exercised inside a `TubularChart`, not only standalone.

- 925 curveframes tests pass
- 118 doctests in the curve-charts guide pass
- clean `rm -rf docs/_build && nox -s docs`, no warnings

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by nstarman** [comment]
Pushed two more commits after finding that the gradient gap I flagged above splits into two very different cases.

## The `d/ds` path was paying an integral it discarded

The custom JVP integrated `dS/dtheta` **unconditionally**. When neither the curve nor `tau_0` is being differentiated, both tangents are symbolically zero, so `dS` is identically zero and that solve was pure waste.

That is not a corner case. Counted on a helix with `s_max`:

| differentiating | `_solve_tau` | dense eval | `_arclength_between` |
|---|---|---|---|
| nothing (forward) | 0 | 1 | 0 |
| `s` alone — **before** | 0 | 1 | **1** |
| `s` alone — **after** | 0 | 1 | **0** |
| curve parameters | 0 | 1 | 1 (correct — still needed) |

`d/ds` covers the chart Jacobian, the induced metric that pulls back through it, and `nearest_tau`'s root-find, all of which carry a fixed curve. Measured: **737,927 us -> 11,292 us, ~65x**, with `|d(gamma)/ds| = 1.000000000000` unchanged and autodiff-vs-finite-difference unchanged at 5.533e-10.

The test is emptiness of `tree_leaves`, the same symbolic-zero check `_tangent_value` already uses. It is a structure question resolved at trace time, so the branch is chosen during tracing and stays jit-safe; a tangent that merely happens to be numerically zero still has a leaf and takes the full path.

## The remaining 1.8x looks irreducible

For gradients with respect to the curve's own parameters, the sensitivity integral cannot be precomputed: `dS/dtheta` depends on which direction in parameter space is being perturbed, and that is only known at the call. A forward-sensitivity precompute (P+1 solves at construction) would fix it for a *fixed* curve — but in a fitting loop theta changes every step, which invalidates it. That is precisely the staleness bug this PR exists to fix, so trading it back would be a regression.

So the honest split, now in the docs as a table rather than one headline number:

| differentiating | cost |
|---|---|
| nothing | a few hundred times cheaper with `s_max` |
| `s` alone | no integration at all |
| curve parameters | about 2x |

Fitting a shape pays the sensitivity integral every step; everything geometric does not.

File: packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py [hunk 1]
```diff
@@ -49,6 +50,18 @@
...
+def _speed(
+    curve: Callable[[Any], Any],
+    tau_unit: u.AbstractUnit,
+    speed_unit: u.AbstractUnit,
+    tau_q: u.AbstractQuantity,
+    /,
+) -> Any:
+    r"""Local speed $\|\boldsymbol{\gamma}'(\tau)\|$ of ``curve`` at ``tau_q``."""
+    dcurve = u.experimental.jacfwd(curve, units=(tau_unit,))
+    return jnp.linalg.norm(dcurve(tau_q).ustrip(speed_unit))
```
[the remark this answers is not in the document]
Comment (nstarman) [reply packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py]: Measured rather than assumed, and declining this one.

I counted the actual invocations on a helix: `_speed` runs **3-4 times per solve**, not once per ODE step. The RHS is traced, so the `jacfwd` closure is built once per trace regardless of how many steps the solver takes:

| | `_speed` calls | `jacfwd` builds |
|---|---|---|
| one eager solve | 3 | 3 |
| dense precompute | 4 | 4 |

So the premise that this scales with step count does not hold, and there is no significant regression to recover.

It is also not a change from this PR: `_speed` is byte-identical to `main`, where `jacfwd` is constructed inside it too. There was no earlier per-solve hoisting for this to defeat.

A micro-optimisation is still available -- 3-4 closure builds is not 1 -- but it is pre-existing and belongs in its own change, not in a PR about gradient correctness.

File: packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py [hunk 2]
```diff
@@ -193,56 +550,89 @@ class ArcLength(eqx.Module):
...
-    `None` is resolved to ``Q(0.0, tau_unit)`` by ``__post_init__``.
+    A `None` passed to `__init__` resolves to ``Q(0.0, tau_unit)``.
     """
 
-    # See `BishopBuilder.diffeqsolver` for why this is static and why the
-    # default is a factory rather than a plain default.
-    diffeqsolver: DiffEqSolver = eqx.field(
-        default_factory=lambda: _DIFFEQSOLVER, static=True
-    )
+    diffeqsolver: DiffEqSolver
```
[the remark this answers is not in the document]
Comment (nstarman) [reply packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py]: Correct, and this was a real regression -- fixed in 290ec8cca.

`main` declares it `eqx.field(default_factory=..., static=True)`; this PR had dropped that on both `ArcLength` and `LagrangianArcLength`. The consequence is measurable:

| | pytree leaves |
|---|---|
| `ArcLength`, before | **12** |
| `BishopBuilder` | 2 |
| `ArcLength`, after | **2** |

The nine extra leaves were exactly what you describe -- `rtol`/`atol` floats, `max_steps` and friends as ints, plus a bool -- so a `jax.tree.map` over the module would have silently rescaled solver tolerances. With `s_max` set it is 11, which is the dense interpolation's genuine array data, not configuration.

Restored on both classes, with the reason recorded on the field.

File: packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py [hunk 3]
```diff
@@ -69,27 +82,331 @@ def _solve_tau(
...
+    not fail just because the search that found it briefly stepped outside
+    while getting there. Only a query genuinely beyond that margin raises,
+    rather than letting `diffrax.DenseInterpolation` clamp internally and
+    return `NaN` silently.
+    """
+    s_unit = s_max.unit
+    s_val = jnp.asarray(s.ustrip(s_unit))
+    s_max_val = jnp.asarray(s_max.ustrip(s_unit))
+
+    slack = _S_MAX_MARGIN * jnp.maximum(1.0, jnp.abs(s_max_val))
```
[the remark this answers is not in the document]
Comment (nstarman) [reply packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py]: Real bug, and fixed in 290ec8cca. Your worked example is exactly right.

`slack = margin * max(1, |s_max|)` stopped shrinking once `s_max` fell below 1, so a 0.001 km domain tolerated a query at 0.05 km -- **fifty times its own length** -- clamped it, and returned the boundary value as though it were the answer.

Now `slack = _S_MAX_MARGIN * jnp.abs(s_max_val)`, a genuine fraction, with a regression test at `s_max = 0.001` that pins both sides: 4% past must pass, 100% past must raise.

On validating `s_max > 0` separately: not added, because it is already loud rather than silent. With a proportional slack, a non-positive `s_max` makes the domain empty, so *every* query raises `_MSG_S_OUT_OF_DOMAIN` immediately. That is a worse-looking error than a dedicated message but not a silent one, and it avoids a construction-time check on a value that may be traced.

I also went after the silence itself and failed, which is worth recording: restricting the slack to the negative side -- tolerate `nearest_tau`'s undershoot, raise on any overshoot -- **breaks the chart round-trip** at `nearest.py:142`. The root-find probes past the *upper* edge too, so the symmetric band is load-bearing. The docstring now states plainly that a caller's own query inside the band is clamped just as silently as a probe is.

File: packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py [hunk 4]
```diff
@@ -172,6 +498,37 @@ class ArcLength(eqx.Module):
...
+    s_max : Quantity, optional
+        When given, the reparametrisation ODE is solved **once**, at
+        construction, as a dense interpolation of $\tau(s)$ over $s \in [0,
+        s_{\max}]$ (`diffrax`'s ``SaveAt(dense=True)``); `__call__` then
+        evaluates that interpolation instead of re-solving. This is a pure
+        performance knob: within $[0, s_{\max}]$, evaluating with and without
+        `s_max` agrees to solver tolerance, and gradients agree too --
+        `__call__`'s reparametrisation always goes through a hand-supplied
+        VJP (`_tau_of_s`) rather than autodiff through the solve or the
+        interpolation, specifically so that swapping the fast path in does
```
[the remark this answers is not in the document]
Comment (nstarman) [reply packages/coordinaxs.curveframes/src/coordinaxs/curveframes/_src/arclength.py]: Right, and there were three of these, not one -- all fixed in 290ec8cca.

- `_solve_tau_dense`: "consumed through `_tau_of_s`'s custom VJP" (the one you flagged)
- `_arclength_between`: "the one extra quadrature `_tau_of_s`'s **backward rule** needs"
- `_arclength_between` again: "one of the two things `_tau_of_s`'s **backward rule** differentiates it against"

All now say JVP.

Worth more than their line count, for the reason you give: this PR's first implementation *was* a `custom_vjp`, and it had to be rewritten when `BishopBuilder`'s forward-mode use of the curve broke it. Docstrings that misname the AD mode are how that class of mistake propagates, so I swept the file for the rest rather than fixing only the flagged line.

**Comment by nstarman** [comment]
Correcting myself: the silent clamp described earlier in this PR is now gone, in 9f9301034. I had written it up as a documented trade-off after one fix attempt failed. That was giving up too early.

## What the clamp cost

`_eval_tau_dense` accepted any `s` within 5% of `[0, s_max]` and clipped it to the boundary. On the guide's own helix with `s_max = 5`:

| query | returned | truth | error |
|---|---|---|---|
| `fast(5.24)` | 1.4367394278 | 1.4942090049 | **6.3e-2**, silent |

## Why it could not simply be deleted

`nearest_tau`'s bracketed root-find genuinely evaluates the curve outside the nominal domain. Instrumenting one chart round-trip, of **163392** queries to `tau(s)`:

- **1016** fall below 0
- **3120** fall above `s_max`
- the extreme is **+/-0.079365079**, i.e. `s_max / 63`

`s_max / 63` is one scan-seed spacing for the default `n_seed = 64`, confirming the docstring's claim to nine digits -- and it is **symmetric**, which is why my earlier idea of keeping slack only on the negative side was refuted: the chart round-trip fails at `nearest.py:142`.

## The fix

Stop deciding who is allowed to overshoot, and make the overshoot region real. `_solve_tau_dense` now integrates `[-m, s_max + m]` instead of `[0, s_max]`, so every query the guard admits is answered from solved data:

| query | before | now |
|---|---|---|
| `s = 5.24` | clamped, wrong by 6.3e-2, silent | agrees with a fresh solve to **~1e-15** |
| `s = -0.24` | clamped to `tau(0)` | correct |
| `s = 5.3` | raises | raises |

Nothing is clipped anywhere. Two solves at construction rather than one -- `tau(0) = tau_0` sits mid-range, so a short backward solve seeds the forward one -- and both are one-off: **per-call cost is unchanged at ~1.6 ms, against ~1.2 s to construct**.

## What made it survive

The old test asserted "does not raise". The clamped, wrong answer passed that. The test now compares against a fresh solve, which is what catches it -- verified by putting the clip back and watching it fail.

One honest consequence: `s = 0` is now read off the interpolation rather than being the ODE's initial condition, so it agrees with a fresh solve to 2e-16 instead of exactly. That is well inside the 1e-10 solver tolerance, and the docstring now says so rather than claiming "no change in behaviour". The full suite caught that where the targeted run did not.

927 tests pass, docs build clean.
